# HG changeset patch
# User Andrew Cooper <andrew.cooper3@citrix.com>
# Date 1377852029 -7200
# Node ID da9f7433c21648d8f608d2d3656a245a6e4c6550
# Parent  49ef1236da05b035345117ac51eca3e363441de7
hvmloader/smbios: Correctly count the number of tables written

Fixes regression indirectly introduced by c/s 4d23036e709627

That changeset added some smbios tables which were option based on the
toolstack providing appropriate xenstore keys.  The do_struct() macro would
unconditionally increment nr_structs, even if a table was not actually
written.

Signed-off-by: Andrew Cooper <andrew.cooper3@citrix.com>
Acked-by: Keir Fraser <keir@xen.org>

diff -r 49ef1236da05 -r da9f7433c216 tools/firmware/hvmloader/smbios.c
--- a/tools/firmware/hvmloader/smbios.c
+++ b/tools/firmware/hvmloader/smbios.c
@@ -192,7 +192,8 @@ write_smbios_tables(void *ep, void *star
 
 #define do_struct(fn) do {                      \
     q = (fn);                                   \
-    (*nr_structs)++;                            \
+    if ( q != p )                               \
+        (*nr_structs)++;                        \
     if ( (q - p) > *max_struct_size )           \
         *max_struct_size = q - p;               \
     p = q;                                      \
